### PR TITLE
Table paging (Draft PR)

### PR DESCRIPTION
**Pull request by rezural**
This adds paging to the table output.

This does change a few methods to return String's instead of printing to screen (this may want to be reworked to be async?), so it's a little invasive. 

still a couple of warnings, and possible errors to be handled correctly. I will look at these soon.

Cheers

### CODE REVIEW

**Comment by fdncred** [comment]
I'm very excited to see this. Hopefully we'll be able to get it whipped into shape soon. Thanks for working on it!

**Comment by rezural** [comment]
There is currently a bug in minus when outputting more than the terminal height at first go, hopefully I will have this fixed today.

https://github.com/arijit79/minus/issues/21

Any thoughts on a pager library preference?

**Comment by jonathandturner** [comment]
@rezural - the main thing is cross-platform support. If you can get that with one crate or if you have to mix and match, we don't have a strong preference.

**Comment by arijit79** [comment]
Hey @jonathandturner I am the author of `minus`. I have fixed one of the two bugs that @rezural mentioned. I will fix the second probably by today

As far as cross compatibility is concerned, I have wrote `minus` with cross compatibility from the ground up. I had originally wrote it for pijul VCS which had the same issue. So I don't think there should be any issues regarding compatibility

**Comment by jonathandturner** [comment]
@arijit79 - that's great to hear! Thanks for the update. Good to know we have a kindred spirit in being cross-platform 👍

**Comment by rezural** [comment]
> I'm very excited to see this. Hopefully we'll be able to get it whipped into shape soon. Thanks for working on it!

Thanks, I think ive come up with a workable solution, hopefully by today something will  be worth looking at.

**Comment by rezural** [comment]
OK,

the last round of changes make it work almost 100%, just need to stop minus clearing the screen unless we have overflowed.

**Comment by rezural** [comment]
There are some test that are failing, I think this is because the paging code clears the screen on small input, not sure really.

but printing directly to screen (instead of utilizing the pager code) makes those tests green again. Perhaps putting this behind a config flag?

**Comment by rezural** [comment]
Also, some lick & spittle needed. I will look at this tomorrow.

Cheers

**Comment by rezural** [comment]
So the failing tests are caused by minus checking if out is a tty, here:

https://github.com/rezural/minus/blob/f50ad353468d4d9e50a6f19cc08542de77627521/src%2Futils.rs#L40

the failing test is here:

https://github.com/nushell/nushell/blob/48a90fea70387181feef1932832d2a07ff900359/tests%2Fshell%2Fpipeline%2Fcommands%2Finternal.rs#L1080

which involves the nu! macro, which is setting up an environment and running commands within..

I see 2 ways to fix this:

check if we are in a tty in the table code
gate this behind a config setting

any thoughts welcome.

**Comment by jonathandturner** [comment]
@andrasio - any idea how we might test this one?

**Comment by rezural** [comment]
I have added a test for tty in the table code, and output via println! if we are not in a tty.

It should work, but crossterm may be causing issues in the CI.

**Comment by rezural** [comment]
So, bringing in minus, means we are bringing in crossterm.

This is not going to build in wasm configuration. I'm at a loss here, and it's something the core team will need to come to a decision one.

**Comment by rezural** [comment]
This is being causes by these function not being defined in crossterm:

https://github.com/crossterm-rs/crossterm/blob/88add302cd8f006704b712c9f6aa624104635c71/src/terminal/sys.rs#L1-L15

This may be a bug in crossterm, or something they may be happy to shim.

**Comment by jonathandturner** [comment]
@rezural - you may want to bring in minus as an optional dependency. We do that with a few crates that won't work in wasm. Then, check that the crate is available before you try to run the pager by using a `cfg` in the code

**Comment by rezural** [comment]
So I have removed this from extras for now, as running the tests with this enabled causes minus to check if it is in a tty, causing errors.

I did have code to test if we were not in a TTY, and disable the pager if this was the case... that seems a bit overkill but I'm happ y to add it back in.

Another option would be to investigate monkey patching using (i.e.) guerilla when testing. but this is also maybe overkill.

Any thoughts?

**Comment by rezural** [comment]
Also note that this depends on #3088. It was originally here, but it made sense to extract those changes out.

**Comment by arijit79** [comment]
@rezural I have published a new version of `minus` with your patches. You should now use the crates.io version rather than the revision

**Comment by rezural** [comment]
To run this, you will have to enable the table-pager feature:

```
 cargo build --features=table-pager && target/debug/nu
```

File: crates/nu-command/src/commands/table/command.rs [hunk 1]
```diff
@@ -187,63 +193,102 @@ async fn table(
...
 
-        start_number += input.len();
+        Result::<_, ShellError>::Ok(())
+    };
+
+    #[cfg(feature = "table-pager")]
+    {
+        let (minus_result, streaming_result) =
+            join(minus::async_std_updating(pager.clone()), stream_data).await;
+        minus_result.map_err(|_| ShellError::untagged_runtime_error("Error paging data"))?;
```
Comment (rezural) [inline crates/nu-command/src/commands/table/command.rs:283]: Is this the correct kind of error to use?

**Comment by jonathandturner** [comment]
I tried this out in Windows with `ls **/*` in the Nushell directory. I got to scroll around the table, which was nice. I tried to exit the pager and then the prompt seems to have frozen.

Do you see the same?

**Comment by rezural** [comment]
OK,

There appears to be some weirdness here. I didn't notice this with ps.

I think this is because the code in table is still streaming data, you will probably notice if you wait long enough, that the prompt returns. Can you try this? For me it takes about 12s to return from ls **/*..

Nice catch :D dammit!

If so I will have a look into adding a variable that the streaming async fn checks, that is set when minus exits. this will require a wrapping async fn, hopefully it is not too painful.

**Comment by jonathandturner** [comment]
@rezural yeah it does return eventually if I wait long enough

**Comment by rezural** [comment]
Ok, this requires using select instead of join.

I also should perhaps be using async_std instead of futures::join.

I'll look at this in an hour or so.

**Comment by rezural** [comment]
@jonathandturner ok that was a bit tricky. But this now observes via a callback when minus has finished, and uses an AtomicBool to check whether we've finished in the stream_data loop. The non-paging code also uses the AtomicBool, otherwise the cfg checking code will start to get a bit silly.

This currently points to a minus branch I havent yet tried to merge with arijit's minus. I will need to look at it again, and make sure that code is sound before creating a pull request on minus.

@arijit79 has been super accomodating, hopefully he will accept my latest changes. but this is starting to really require features from minus, that may or not be in scope of minus. we shall see.

**Comment by rezural** [comment]
You will need to do a fresh checkout. I have simplified the history quite a bit, and removed the Cargo.lock updates.

**Comment by rezural** [comment]
I'm getting segfaults unfortunately now.

Back to debugging.

**Comment by rezural** [comment]
So, while I have gotten a few segfaults, this is very intermittent.

Probably needs some more people + architectures to see how flaky the code is.

I have a couple of ideas where it might be, but am a little out of my depth.

**Comment by rezural** [comment]
I have only seen segfaults while running 

```
ls**/* 
```

and holding PgDn, 

within the nu source directory, in debug mode. If anyone is willing to test this, it would be much appreciated.

**Comment by rezural** [comment]
Segfaults are fixed in upstream minus.

This is currently pointing to my branch of minus, until those changes are merged (hopefully), and another branch is merged (add-finished-callback)

**Comment by jonathandturner** [comment]
This is look good!

I played around with it and everything seemed to work well.

@rezural I was wondering, do we want to change the "minus" text at the bottom of the screen? Doesn't have to be this PR but figured we might want to replace it with some instructions or something.

**Comment by jonathandturner** [comment]
Another one we might also want to improve in a follow-up PR: If you do `/` to do a search, you can search but it won't jump to the first result. That'd be a nice thing to add at some point.

**Review by rezural — COMMENTED** [review]
> Another one we might also want to improve in a follow-up PR: If you do `/` to do a search, you can search but it won't jump to the first result. That'd be a nice thing to add at some point.

Yep, agreed. This is something that will have to be done on the minus side... I'm starting to feel bad for all the PR requests piling up!

**Comment by jonathandturner** [comment]
Looks good on Windows, too

**Comment by rezural** [comment]
> This is look good!
> 
> I played around with it and everything seemed to work well.
> 
> @rezural I was wondering, do we want to change the "minus" text at the bottom of the screen? Doesn't have to be this PR but figured we might want to replace it with some instructions or something.

Of course, I meant to mention this, but it's such tiny fix, that it got swamped by all the other stuff. trivial to do.

File: crates/nu-command/src/commands/table/command.rs [hunk 2]
```diff
@@ -187,63 +193,102 @@ async fn table(
...
+        streaming_result?;
     }
 
+    #[cfg(not(feature = "table-pager"))]
+    stream_data
+        .await
+        .map_err(|_| ShellError::untagged_runtime_error("Error streaming data"))?;
+
```
[the remark this answers is not in the document]
Comment (jonathandturner) [reply crates/nu-command/src/commands/table/command.rs:312]: I think it's okay here. Often people will see then as a result of an `autoview` they didn't type, so there isn't a good place to put the underline for the error.

**Comment by jonathandturner** [comment]
Maybe something like:

"Arrow keys/PgUp/PgDown to move. 'q' to quit"

**Comment by rezural** [comment]
> Maybe something like:
> 
> "Arrow keys/PgUp/PgDown to move. 'q' to quit"

Yes, the keybindings are a little weird. It would be good to have this settable by client code.

The current keybindings are here, but for nushell, I could perhaps fork?

https://github.com/arijit79/minus#end-user-help

**Comment by rezural** [comment]
But, yeah. 

I am definately starting to hit the conceptual edges of what minus offers. I will have a chat with arijit about changes he would be willing to accept. I think he is on another project right now, so probably doesnt have many cycles to merge too many changes.

**Comment by jonathandturner** [comment]
@rezural - if arijit is interested in changes, might be good to see if we can do that. If not, yeah we can talk about forking. We've done that with a few crates that Nushell needs.

**Comment by fdncred** [comment]
i haven't really been following too closely but i checked it out today with `ls **/*` on windows and there's no paging of about ~1000 files. Is there some config thing to turn it on?

**Comment by jonathandturner** [comment]
@fdncred you need to build with the feature enabled (it's behind a feature flag for now)

```
cargo run --features=table-pager
```

**Comment by fdncred** [comment]
i think i answered my own question it's behind `--features=table-pager` - building now. :)

**Comment by fdncred** [comment]
@rezural wow! a lot has changed since i tried last, i.e. it works. haha!  Great job. a few things are driving me crazy (keybindings) but generally i love it. I would like to be able to invoke it instead of making paging the default. I like when the table is left on the screen and i can scroll up and review it. with paging, once you hit q, it's gone.

**Comment by jonathandturner** [comment]
I wonder if it would be possible to turn it into our own version of `less`. We can do that in separate steps, like land this behind a feature flag, then move it to being a command.

Then we could also add a config setting that let you make it the default.

**Comment by fdncred** [comment]
I like where you're going with that @jonathandturner. I'd support all those ideas!

**Comment by rezural** [comment]
> @rezural wow! a lot has changed since i tried last, i.e. it works. haha! Great job. a few things are driving me crazy (keybindings) but generally i love it. I would like to be able to invoke it instead of making paging the default. I like when the table is left on the screen and i can scroll up and review it. with paging, once you hit q, it's gone.

Thanks.

Yeah, I don't personally like the keybindings. I have some ideas to make minus more amenable to arbitrary key bindings, hopefully in the next few days. perhaps in this PR, or perhaps in a separate one. It would be good to get it in right, so you guys dont get a bunch of questions/complaints about the keybindings on the discord. inevitable that one.

Fair point on making it explicit. I'm not sure that having it in table is actually the right place. Now I've worked out all the issues, it should be fairly seamless to put it in a more appropriate place. I'm not sure how the table string data would get passed into, say a less command with how the pipeline currently works (Untagged string?), but I've yet to look properly at nu's data structures for structured data..

**Comment by jonathandturner** [comment]
@rezural - I think I can make the passing the text in work as well.

@fdncred - we could also use this to do text viewing, too. That way we don't have to ship less.exe any more. We'd just have all the paging stuff built in and could pull it up.

**Comment by jonathandturner** [comment]
@rezural - I was chatting with @fdncred and we agreed this is probably ready to land now, and we can continue improving it with additional PRs after that to get it where we want. Landing it lets other people try it out and give feedback a little easier, which will help us improve it further.

**Comment by rezural** [comment]
@jonathandturner Absolutely fine with me! I will look at the keybinding code in the next couple of days, and submit a PR.

Cheers

**Comment by andrasio** [comment]
> @andrasio - any idea how we might test this one?

Just noticed this :(

**Pull request merged**